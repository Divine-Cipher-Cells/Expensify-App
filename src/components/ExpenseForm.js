import React from "react";
import moment from "moment";
import { SingleDatePicker } from "react-dates";


//const now = moment();
//console.log(now.format("MMMM YYYY Do a"));


export default class ExpenseForm extends React.Component{
    constructor(props){
        super(props);
        this.state ={
            description: props.expense ? props.expense.description :"",
            note: props.expense ? props.expense.note :"",
            amount: props.expense ? (props.expense.amount/100).toString() :"",
            createdAt: props.expense ? moment(props.expense.createdAt) : moment(),
            focused: false,
            error:""
        };

    }


    onDescriptionChange = (e) => {
        const description = e.target.value;
        this.setState( () =>({
            description
        }));
    };

    onNoteChange = (e) =>{
        const note = e.target.value;
        this.setState( () => ({
            note
        }));
    };

        onAmountChange = (e) => {
            const amount= e.target.value;
            if(!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)){
                this.setState(() => ({
                    amount
                }));
            }
        };
        onDateChanging = (createdAt) => {
            if (!!createdAt)
            this.setState( () =>({createdAt}));
        };
        onFocusChange =({focused}) => {
            this.setState( () => ({focused}));
        };

        onSubmitForm = (e) => {
            e.preventDefault();
            const error = <div>Please fill the description and form</div>
            if(!this.state.description || !this.state.amount){
                this.setState( () => ({
                    error
                }));
            }else {
                this.setState(() =>({
                    error:""
                }));
                this.props.whenSubmit(this.state);
            }
        };
        
    
    render(){
            return (<div>
                <p>{this.state.error}</p>
                <form onSubmit={this.onSubmitForm}>
                <input type="text" placeholder="Description" autoFocus value={this.state.description} onChange = {this.onDescriptionChange} />
                <input type="number" placeholder="Amount" value={this.state.amount} onChange = {this.onAmountChange} />
                <SingleDatePicker date={this.state.createdAt} onDateChange={this.onDateChanging} 
                focused={this.state.focused} onFocusChange={this.onFocusChange}
                numberOfMonths={1}
                isOutsideRange={() => false}
                />
                <textarea placeholder="Add a note for your expense (optional)" value={this.state.note} onChange={this.onNoteChange}></textarea>
                <button>Add Expense</button>
                </form>
                </div>);
};
}

// here we track the updation and filling of form by local state