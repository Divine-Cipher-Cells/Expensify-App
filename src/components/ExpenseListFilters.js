import React from "react";
import { DateRangePicker } from "react-dates";
import { connect } from "react-redux";
import { setStartDate, setTextFilter, sortByAmount ,sortByDate,setEndDate} from "../actions/filters_actions";

export class ExpenseListFilters extends React.Component{
    constructor(props){
        super(props);
        this.state = {
            calendarFocused: null
        };
    }
    changeDualDates = ({startDate,endDate}) => {
        this.props.setStartDate(startDate);
        this.props.setEndDate(endDate);
    };

    onFocusChange = (calendarFocused) => {
        this.setState( () => ({ calendarFocused }));
    };
    ontextFilterChange =(e) =>{
        const inputValue= e.target.value;
        this.props.setTextFilter(inputValue);
    };

    onSelectValueChange= (e) => {
        e.target.value === "date" ? this.props.sortByDate() : this.props.sortByAmount() ;
    };

render(){
return (    <div>
    <input type="text" value={this.props.filters.text} onChange={this.ontextFilterChange} />
    <select value={this.props.filters.sortBy} onChange={this.onSelectValueChange}>
    <option value="date">Date</option>
    <option value="amount">Amount</option>
    </select>
    <DateRangePicker focusedInput={this.state.calendarFocused} startDate={this.props.filters.startDate} endDate={this.props.filters.endDate} 
    onDatesChange= {this.changeDualDates} onFocusChange={this.onFocusChange}
    numberOfMonths={1}
    isOutsideRange={() => false}
    showClearDates={true}
    />
    </div>);
}
}

const mapStateToProps = (state,props) =>({
    filters: state.filters
});

const mapDispatchToProps = (dispatch,props) =>({
setTextFilter: (inputValue) => dispatch(setTextFilter(inputValue)),
setStartDate:(startDate) => dispatch(setStartDate(startDate)),
setEndDate:(endDate) => dispatch(setEndDate(endDate)),
sortByAmount:() => dispatch(sortByAmount()),
sortByDate:() => dispatch(sortByDate())
});

export default connect(mapStateToProps,mapDispatchToProps)(ExpenseListFilters);

//using value we set the form element to only read only value
// to read and update it we need to first let it access store 
// by accessing store it also gets a dispatch() method as dispatch prop