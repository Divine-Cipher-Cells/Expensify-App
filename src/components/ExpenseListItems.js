
import React from "react";

import { Link } from "react-router-dom";

const ExpenseListItems = ({ id,description, amount, createdAt }) => (
    <div>
    <Link to={`/edit/${id}`}> <h3>{description}</h3></Link>
   
    <p>{amount} ~ {createdAt}</p>
    </div>
  
);

export default ExpenseListItems;
// here i an not using mapStateToProps as the button doesn't need it to run this
// here we access dispatch via expense 