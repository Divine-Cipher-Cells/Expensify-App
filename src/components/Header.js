import React from "react";
import { NavLink } from "react-router-dom";
const Header = (props) =>(
    <header>
    <h1>{props.children}</h1>
    <NavLink to="/" activeClassName="is-active" exact strict >Dashboard </NavLink>
    <NavLink to="/create" activeClassName="is-active" exact strict>Create-Expense </NavLink>
    </header>
);
export default Header;