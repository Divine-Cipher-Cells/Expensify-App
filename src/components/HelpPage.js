import React from "react";
const HelpPage = (props) => (
    <div>
    This is from my help component.
    </div>
);
export default HelpPage;