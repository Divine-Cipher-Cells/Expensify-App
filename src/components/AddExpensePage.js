import React from "React";
import { connect } from "react-redux";
import ExpenseForm from "./ExpenseForm";
import {addExpense} from "../actions/expenses_actions";

export class AddExpensePage extends React.Component{
    whenSubmit = (expense) => {
        this.props.addExpense(expense);
        this.props.history.push("/");

    };

    render(){
        return (
            <div>
            <h1>Add Expense</h1>
            <ExpenseForm whenSubmit = {this.whenSubmit}/>
            </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
    addExpense:(expense) => dispatch(addExpense({
        description: expense.description,
    amount: parseFloat(expense.amount,10)*100,
    createdAt: expense.createdAt.valueOf(),
    note: expense.note
    })),
});
export default connect(undefined,mapDispatchToProps)(AddExpensePage);


// note we need to parse number from string to a float value(and also remove the decimals as we are taking the value in pennies)
// and also need to parse createdAt as it will give an improper data and thus we need data in miliseconds so we 
// one of the moment apis to convert present state to milliseconds this api is valueOf() method