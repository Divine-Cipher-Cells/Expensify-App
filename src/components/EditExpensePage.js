import React from "react";
import {connect} from "react-redux";
import { editExpense } from "../actions/expenses_actions";
import { removeExpense } from "../actions/expenses_actions";
import ExpenseForm from "./ExpenseForm";

 export class EditExpensePage extends React.Component{
    whenSubmit = (expense) => {
        this.props.editExpense(this.props.expense.id,expense);
        this.props.history.push("/");

    };
    removeExpense = () => {
        this.props.removeExpense({id:this.props.expense.id});
        this.props.history.push("/");
    };


    render(){
        return(
            <div>
            <ExpenseForm expense={this.props.expense} whenSubmit= {this.whenSubmit} />
            <button onClick={this.removeExpense}>REMOVE</button>
            </div>
        );
    }
}

const mapStateToProps =(state, props) =>({
expense: state.expenses.find( (expense) => expense.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
editExpense: (id,expense) => dispatch(editExpense(id,{
    description: expense.description,
amount: parseFloat(expense.amount,10)*100,
createdAt: expense.createdAt.valueOf(),
note: expense.note
})),

removeExpense: (expense) => dispatch(removeExpense({id: expense.id}))
});
export default connect(mapStateToProps, mapDispatchToProps)(EditExpensePage);



//       props.dispatch(editExpense(props.expense.id, {
    //   description: expense.description,
    //   amount: parseFloat(expense.amount,10)*100,
    //   createdAt: expense.createdAt.valueOf(),
    //   note: expense.note
    //   }));
    //   we here simplify this using mapDispatchToProps


    // here we do not use second argument props in mapdispatch as it could not provide mapstate props