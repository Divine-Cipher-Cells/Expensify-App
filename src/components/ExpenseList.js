import React from "react";
import { connect } from "react-redux"; 
import getVisibleExpenses from "../selectors/expenses_selector";
import ExpenseListItems from "./ExpenseListItems";

export const ExpenseList = (props) => (
<div>

<h2>Expense List Items</h2>
{props.expenses.length === 0 ? <p>No expense</p> : props.expenses.map( (expense,index) => {
    return <ExpenseListItems key={expense.id} {...expense} />;
})}
</div>
);
const mapStateToProps = (state) =>( {
    expenses: getVisibleExpenses(state.expenses,state.filters),
    });

export default connect(mapStateToProps)(ExpenseList);
 // store state gets passed inside the inside funxtion of connect

 // we reference something when we don t want return value but the function to execute the conditions
 // as we do not want to use the return value here just want to get it as a props for the wrapped component so we only reference it

 // NOte before EXpenseListener was not conected to store but now it is so it has now both the props as supplied from here and also the ones it got from store