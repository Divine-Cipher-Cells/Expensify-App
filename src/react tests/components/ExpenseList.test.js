import { ExpenseList } from "../../components/ExpenseList";
import React from "react";
import { shallow } from "enzyme";
import expenses from "../fixtures/expenses_fixtures";

test("ExpenseList component",() => {
const wrapper = shallow(<ExpenseList expenses={expenses} />);
expect(wrapper).toMatchSnapshot();
});

test("should render expense list with empty message", () =>{
    const wrapper = shallow(<ExpenseList expenses={[]} />);
    expect(wrapper).toMatchSnapshot();
});