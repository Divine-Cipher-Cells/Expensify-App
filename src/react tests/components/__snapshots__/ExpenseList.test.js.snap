// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ExpenseList component 1`] = `
<div>
  <h2>
    Expense List Items
  </h2>
  <ExpenseListItems
    amount={195}
    createdAt={345600000}
    description="Gum"
    id="1"
    key="1"
    note=""
  />
  <ExpenseListItems
    amount={109500}
    createdAt={-345600000}
    description="Rent"
    id="2"
    key="2"
    note=""
  />
  <ExpenseListItems
    amount={4500}
    createdAt={0}
    description="Credit Card"
    id="3"
    key="3"
    note=""
  />
</div>
`;

exports[`should render expense list with empty message 1`] = `
<div>
  <h2>
    Expense List Items
  </h2>
  <p>
    No expense
  </p>
</div>
`;
